### ISSUE TITLE
Install instructions for Linux are missing in Readme

### ISSUE THREAD

**Issue by driqeks** [lllyasviel/stable-diffusion-webui-forge#138]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits

### What would your feature do ?

1. Install instructions for Linux are missing in Readme
2. There also is no information at all in the readme that downloading the "[>>> Click Here to Download One-Click Package<<<]" is not working on Linux. I downloaded it and then was surprised that it only includes .bat files and no .sh files.

### Proposed workflow

1. Go to the Readme
2. See steps how to install on Linux

**Comment by AlexNolasco** [comment]
I was looking for the same, so I ended up with  this
```
python -m venv .venv
. ./venv/bin/activate
pip install -r requirements.txt
. ./webui.sh --xformers --api
```

**Comment by Yumae** [comment]
> I was looking for the same, so I ended up with this
> 
> ```
> python -m venv .venv
> . ./venv/bin/activate
> pip install -r requirements.txt
> . ./webui.sh --xformers --api
> ```

This seems to pull in dependencies for NVIDIA cards only. How can i make it download the ROCm equivalents to run it on AMD?

**Comment by AlexNolasco** [comment]
> > I was looking for the same, so I ended up with this
> > ```
> > python -m venv .venv
> > . ./venv/bin/activate
> > pip install -r requirements.txt
> > . ./webui.sh --xformers --api
> > ```
> 
> This seems to pull in dependencies for NVIDIA cards only. How can i make it download the ROCm equivalents to run it on AMD?

I would check if https://pinokio.computer/ has support for it

**Comment by epictetito** [comment]
> I was looking for the same, so I ended up with this
> 
> ```
> python3 -m venv venv
> . ./venv/bin/activate
> pip install -r requirements.txt
> . ./webui.sh --xformers --api
> ```
man, can you be a little more explicit?

I understand that you download and unpack webui_forge_cu121_torch21.7z

What do you do next?

do you then execute those commands in some specific directory?

can you explain what are the 2 dots in ". ./venv/bin/activate"'?

Thanks

**Comment by epictetito** [comment]
> I was looking for the same, so I ended up with this
> 
> ```
> python3 -m venv venv
> . ./venv/bin/activate
> pip install -r requirements.txt
> . ./webui.sh --xformers --api
> ```

I managed to install it thanks to you. THanks, but the performance with XL models is much worse than on A1111, with 1.5 models practically the same. is your experience the same?. NVIDIA GeForce RTX 3060

**Comment by PaPa1ya** [comment]
Followed [these](https://www.reddit.com/r/StableDiffusion/comments/zu9w40/novices_guide_to_automatic1111_on_linux_with_amd/) instructions to work on ubuntu, rocm5.7, rx6650xt. In principle it does not differ much from the official one from a1111.

**Comment by DWeiers** [comment]
Just in case someone is stumbling across this and doesn't get it to work:

I don't know if it was differrent before, but at the time of this post installation on linux is actually pretty simple.

## Requirements
Make sure you have python 3.10 and git installed on your machine.

## Installation
1. In terminal go to whatever directory you want the installation to be in.
3. Clone the repository using this command:
`git clone https://github.com/lllyasviel/stable-diffusion-webui-forge`
This will create a directory named _stable-diffusion-webui-forge_ and download the repository to it.
4. Go into this newly created directory:
`cd stable-diffusion-webui-forge`
5. Start the webui by typing:
`./webui.sh`
This will start the local webui server and open it on `http://127.0.0.1:7861` in the standard browser.

During the first startup the script will download and install some dependencies, be patient, this might take a while (depending on your machine and your internet connection).

## Starting Forge UI
To start the webui simply launch _webui.sh_ in the installation directory (_stable-diffusion-webui-forge_).

## Updating
To update Forge UI simply follow these steps:
1. Make sure forge is not running.
2. In terminal go **into** the installation directory (_stable-diffusion-webui-forge_)
3. Run the following command:
`git pull origin main`
This will repull the newest version of the repository.

**Comment by ngocdaothanh** [comment]
> ...
> ./webui.sh

Thanks @DWeiers :+1:
 
It's Jan 8 2025 now and I confirm that your instruction works on Pop OS (this distro is based on Ubuntu and it has builtin support for Nvidia).

**Comment by Meduty** [comment]
Hi! 

Does this method of installing on windows handle CUDA/Torch correctly?

**Comment by DWeiers** [comment]
As far as I remenber, it did. The script should install everything on its first run. Its been some time, though. So I am not sure. Just give it a try. If it doesnt you can just delete the folder.

**Comment by dchawisher** [comment]
> As far as I remenber, it did. The script should install everything on its first run. Its been some time, though. So I am not sure. Just give it a try. If it doesnt you can just delete the folder.

This is great, but how do we use command-line flags? Webui-user.sh seems to do nothing and it is not obvious what I need to uncomment to make it do something. (Recently switched from Windows so sorry if that is a dumb question.)

**Comment by DWeiers** [comment]
Hi @dchawisher 
As far as I remember I never used any flags. If the script webui.sh doesn't accept any parameters I don't know. Sorry, but tbh I only used forge a few times before switching to invoke.

**Comment by lalululapont** [comment]
I have tried everithing but cant find the way to install this on ubuntu.
when putting "./webui.sh" said:
ERROR: Cannot activate python venv, aborting...

**Comment by az14el** [comment]
> I have tried everithing but cant find the way to install this on ubuntu. when putting "./webui.sh" said: ERROR: Cannot activate python venv, aborting...

same results here, popOS 22.04

**Comment by az14el** [comment]
> > I have tried everithing but cant find the way to install this on ubuntu. when putting "./webui.sh" said: ERROR: Cannot activate python venv, aborting...
> 
> same results here, popOS 22.04

after trying "python3 -m venv venv/ --without-pip" (the above error was an ensurepip issue) i was able to get the activate files in venv/bin, but now run into an issue with installing torch

Installing torch and torchvision
/home/user/stable-diffusion-webui-forge/venv/bin/python: No module named pip
Traceback (most recent call last):
  File "/home/user/stable-diffusion-webui-forge/launch.py", line 54, in <module>
    main()
  File "/home/user/stable-diffusion-webui-forge/launch.py", line 42, in main
    prepare_environment()
  File "/home/user/stable-diffusion-webui-forge/modules/launch_utils.py", line 428, in prepare_environment
    run(f'"{python}" -m {torch_command}', "Installing torch and torchvision", "Couldn't install torch", live=True)
  File "/home/user/stable-diffusion-webui-forge/modules/launch_utils.py", line 125, in run
    raise RuntimeError("\n".join(error_bits))
RuntimeError: Couldn't install torch.
Command: "/home/user/stable-diffusion-webui-forge/venv/bin/python" -m pip install torch==2.3.1 torchvision==0.18.1 --extra-index-url https://download.pytorch.org/whl/cu121
Error code: 1

i see that forge was tested with debian and fedora, im not 100% attached to popOS, would these be more recommended or is this more just a simple user error thing? (i am windows smoothbrain for last 10 years and have only used the 1 click portable install)

**Comment by DWeiers** [comment]
Well I'm not an expert but as far as I get it the problem seems to be that the forge installer needs pip for installing components. pip is very common for installing python components. It looks like you set up your virtual environment without pip. Therefore the installer can't find pip which is necessary to install (in this case) torch and torchvision. 

I don't think this is distribution dependent. It's rather an issue with how the virtual environment was set up.
AFAIK popOS is based on ubuntu, so it should work fine.

I just tested it on tuxedoOS (which is based on ubuntu as well). The installer (the `webui.sh` script) created a virtual environment on its own. It even installed the latest version of pip (since mine was apparently outdated). So there shouldn't be the need for any python magic on your side.

You have to make sure that python is installed, though. You can check this with `python3 --version` it should state 3.10 or above. I have 3.11.11 which worked fine for me.

In case you don't have python installed you need to install it first. Debian and the vast majority of distros derived from it (which includes Ubuntu and almost all of the ubuntu based distros; yes, AFAIK that does include popOS) usually use `apt` for package management (installation of programs). On those distros you can install python3 by issuing the following commands:

`sudo apt update`
`sudo apt install python3`

If python is installed but pip isn't you can install pip with

`sudo apt update`
`sudo apt install python3-pip`

Other distros might not use `apt` but other package management tools. Fedora for example uses `dnf`. The usage is similar though. The command to install python on fedora should be:

`sudo dnf install python3`

Basically you have to check which package-manager your distro uses and then use it to install `python3`. `pip`should be included in a standard python install, but if for whatever reason it isn't you need to install `python3-pip` as well.

Hope that helps.

**Comment by az14el** [comment]
thanks heaps for that, confirmed than python 3.10 and python3-venv were included but if starting from step 1 again still doesn't work I'll try installing pip myself (chat GPT seems to think popOS 22.04 includes pip but i suspect it doesn't given where i got stuck, but benefit of the doubt just in case i did something wrong or out of order), will update this reply then

*update
-used sudo apt update
-then sudo apt install python3-pip, then retried webui.sh
which led to the same "cannot activate python venv, aborting" error
realised i was mistaken about popOS having python3-venv (or at least an up to date one? it could create a virtual environment folder by both terminal command and through webui.sh)
-so tried sudo apt install python3-venv (i think that was the command used, was the command recommended by webui.sh after all that)
-running webui.sh then correctly installed all dependencies without issue and was able to open the UI, time to swap over some models and make sure all is well, but seems so, thanks again, came down to plain old unfamiliarity with the OS afterall

*update 2
working as intended, hope this helps someone else having issues with ubuntu based distros, they could just not have a complete or up to date python package, in popOS 22.04's case it needed both pip & venv installed via terminal
